File: CppBaseWrapper/src/communication/kafka/KafkaConsumer.cpp
/*******************************************************************************
 * Copyright 2021 Moritz Gütlein
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
#include "KafkaConsumer.h"
using namespace daceDS;
std::function<void()> on_signal;

void signal_handler(int) { on_signal(); }
void KafkaConsumer::subscribe(std::string broker,
                              std::vector<std::string> &topics,
                              std::string group_id, std::string loglevel) {

    group_id = Config::getInstance()->getScenarioID()+"_"+Config::getInstance()->getSimulatorID()+"_"+group_id;
    KDBGCB("Group id is " << group_id);
    KDBGCB("Brokers: " << broker);

    conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
    tconf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);

    std::string errstr = "";
    tconf->set("auto.offset.reset", "earliest", errstr);
    tconf->set("topic.auto.offset.reset", "earliest", errstr);
    //tconf->set("enable.auto.commit", "true", errstr);
    tconf->set("request.required.acks", "1", errstr);

    conf->set("metadata.broker.list", broker, errstr);
    conf->set("default_topic_conf", tconf, errstr);
    conf->set("enable.partition.eof", "false", errstr);
    conf->set("group.id", group_id, errstr);
    conf->set("client.id", group_id, errstr);
    conf->set("acks", "1", errstr);
    conf->set("message.send.max.retries", "10000000", errstr);
    conf->set("max.in.flight.requests.per.connection", "1", errstr); //important for correct ordering
    conf->set("enable.auto.commit", "true", errstr);

    conf->set("topic.metadata.refresh.interval.ms", "1000", errstr);
    conf->set("metadata.max.age.ms", "1000", errstr);
    conf->set("allow.auto.create.topics", "true", errstr);
    conf->set("auto.offset.reset", "earliest", errstr);
    conf->set("max.partition.fetch.bytes", "10000000", errstr);
    conf->set("socket.blocking.max.ms", "6000", errstr);
    conf->set("fetch.min.bytes", "1", errstr); //latency
    conf->set("fetch.wait.max.ms", "1", errstr); //latency?
    //conf->set("debug", "all", errstr);

    // consumer = RdKafka::Consumer::create(conf, errstr);

    conf->set("default_topic_conf", tconf, errstr);
    KDBGCB("Created config " << errstr);

    consumer = RdKafka::KafkaConsumer::create(conf, errstr);

    if (!consumer) {
        KERROR("Failed to create consumer: " << errstr);
        exit(1);
    }

    KDBGCB("Created consumer " << consumer->name());
    //conf->set("dr_cb", &ex_dr_cb, errstr);
    //queue = RdKafka::Queue::create(consumer);

    KDBGCB("topic list contains number of entries: "<< topics.size());
    for (auto topicStr : topics) {
        KDBGCB("Subscribing to " << topicStr);
        //     RdKafka::Topic *topic = RdKafka::Topic::create(consumer,
        //                                                    topicStr,
        //                                                    tconf,
        //                                                    errstr);
        //     if (!topic) {
        //         KERROR("Failed to create topic: " << errstr);
        //         exit(1);
        //     }
        //     int32_t partition = 0;
        //     int64_t start_offset = RdKafka::Topic::OFFSET_BEGINNING;
        //     //RdKafka::ErrorCode resp = consumer->start(topic, partition, start_offset, queue);
        //     RdKafka::ErrorCode resp = consumer->subscribe(topic, partition, start_offset, queue);
        //     if (resp != RdKafka::ERR_NO_ERROR) {
        //         KERROR("Failed to start consumer: " << RdKafka::err2str(resp));
        //         exit(1);
        //     }
    }
    RdKafka::ErrorCode resp = consumer->subscribe(topics);
    _topics = topics;

    if (resp != RdKafka::ERR_NO_ERROR) {
        KERROR("Failed to start consumer: " << RdKafka::err2str(resp));
        exit(1);
    }

    Consumer::subscribe(broker, topics, group_id, loglevel);

    // on_signal = [&]() { KDEBUG("on_signal"); };
    // signal(SIGINT, signal_handler);
    // signal(SIGABRT, signal_handler);
    KDBGCB("Starting listener thread");
    t1 = std::thread(&KafkaConsumer::listen, this);
    //t1.detach();
}

void KafkaConsumer::msg_consume(RdKafka::Message *message, void *opaque) {
    switch (message->err()) {
        case RdKafka::ERR__TIMED_OUT:
            break;

        case RdKafka::ERR_NO_ERROR:
            /* Real message */
            KDBGCB("Received on topic="
                   << message->topic_name() << " @ time="
                   << message->timestamp().timestamp << " @ offset="
                   << message->offset());
            {
                try {
                    std::string key = "";
                    std::string sender = "";
                    int64_t st = 0;
                    
                    RdKafka::Headers* headers = message->headers();
                    if (headers) {
                        int64_t time = -1;
                        uint32_t epoch = -1;
                        std::vector<RdKafka::Headers::Header> hdrs = headers->get_all();
                        for (size_t i = 0; i < hdrs.size(); i++) {
                            const RdKafka::Headers::Header hdr = hdrs[i];

                            if (hdr.value() != NULL){
                                if(hdr.key() == "sender"){
                                    sender =  static_cast<const char *>(hdr.value());

                                    //is this our own message? -> ignore it
                                    if(sender.find(Config::getInstance()->getSimulatorID())  != std::string::npos ){
                                        KDBGCB("ignoring own message! sender=" << sender<< ", own id="<<Config::getInstance()->getSimulatorID());
                                        consumer->commitSync();
                                        return;
                                    } 
                                    else {
                                        KDBGCB("accepting message! sender=" << sender << ", own id="<<Config::getInstance()->getSimulatorID());
                                    }
                                }
                                else if(hdr.key() == "time"){
                                    // time = *reinterpret_cast<const int64_t*>(hdr.value());     
                                    const uint8_t *array = reinterpret_cast<const uint8_t *>(hdr.value());
                                    KDBGCB("received time: " << std::hex << (int) array[0] << (int) array[1]<< (int) array[2] << (int) array[3]<< (int) array[4] << (int) array[5]<< (int) array[6] << (int) array[7] << std::dec);

                                    time = 
                                        static_cast<int64_t>(array[7]) |
                                        static_cast<int64_t>(array[6]) << 8 |
                                        static_cast<int64_t>(array[5]) << 16 |
                                        static_cast<int64_t>(array[4]) << 24 |
                                        static_cast<int64_t>(array[3]) << 32 |
                                        static_cast<int64_t>(array[2]) << 40 |
                                        static_cast<int64_t>(array[1]) << 48 |
                                        static_cast<int64_t>(array[0]) << 56;          
                                    KDBGCB("decoded to " << time) ;
                                }
                                else if(hdr.key() == "epoch"){
                                    // epoch = *reinterpret_cast<const int*>(hdr.value()); 
                                    const uint8_t *array = reinterpret_cast<const uint8_t *>(hdr.value());                                    
                                    KDBGCB("received epoch: " << std::hex << (int) array[0] << (int) array[1]<< (int) array[2] << (int) array[3]<< std::dec);

                                    epoch = 
                                        static_cast<uint32_t>(array[3]) << 0 |
                                        static_cast<uint32_t>(array[2]) << 8 |
                                        static_cast<uint32_t>(array[1]) << 16 |
                                        static_cast<uint32_t>(array[0]) << 24;     
                                    KDBGCB("decoded to " << epoch) ;
                                }
                            }
                        }

                        if (message->key()) {
                            key = *message->key();
                        }
                


                        if(time < 0 || epoch < 0){
                            KDBGCB ("time or epoch of received msg below 0");
                            // exit(1);
                        } else {
                            st =  time*1000+epoch;
                        }
                        KDBGCB("HEADERS: sender=" << sender << ", time="<<time<<", epoch="<< epoch<<", st="<< st);
                        

                    } else { 
                        KERROR ("no headers in msg on topic "<< message->topic_name());
                    }

                    ConsumedMessage* cmsg = new ConsumedMessage(message->topic_name(),
                                                                key, 
                                                                (message->timestamp()).timestamp,
                                                                sender,
                                                                message->len(),
                                                                st,
                                                                message->payload());
                    bool p = process(cmsg);
                } 
                catch(std::exception &e)  {
                    KERROR("msg_consume ERROR: " << e.what());
                }
            }
            consumer->commitSync();

            break;

        case RdKafka::ERR__PARTITION_EOF:
            /* Last message */
            KDBGCB("Consumed EOF MSG");
            if (exit_eof) {
                run = 0;
            } else{
                KDBGCB("--> Currently ignoring EOF MSG");
            }
            break;

        case RdKafka::ERR__UNKNOWN_TOPIC:
        case RdKafka::ERR__UNKNOWN_PARTITION:
        case RdKafka::ERR_UNKNOWN_TOPIC_OR_PART:
            KERROR(message->errstr());
            //run = 0;
            //consumer->subscribe(_topics);
            break;

        default:
            /* Errors */
            KERROR("Consume failed: " << std::to_string(message->err()) << " " << message->errstr());
            //consumer->subscribe(_topics);
            exit(1);
    }
}

void KafkaConsumer::listen() {
    KDBGCB("Starting to listen on topics: ");
    for (auto t : _topics) {
        KDEBUG(t);
    }

    usleep(1000 * 1000);
    while (run) {
        // RdKafka::Message *msg = consumer->consume(queue, 1000);
        try {
            RdKafka::Message *msg = consumer->consume(10);

            msg_consume(msg, NULL);
            delete msg;

        }  catch  (std::exception &e)  {
            KERROR(e.what());
        }

        //consumer->poll(1); //poll should never be used with RdKafka::KafkaConsumer
    }

    KDBGCB("Finished listening on topics: ");
    for (auto t : _topics) {
        KDEBUG(t);
    }
}

void KafkaConsumer::stop() {
    run = 0;
    t1.join();
    consumer->close();
}

File: PythonBaseWrapper/src/communication/KafkaConsumer.py
#!/usr/bin/env python3
#
# Copyright 2021 Moritz Gütlein
# 
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
# 

from confluent_kafka import avro
import csv
from confluent_kafka.avro.serializer import SerializerError
import json
from confluent_kafka import KafkaError
from AvroProducerConsumer import AvroConsumerStrKey
from confluent_kafka import Consumer
import threading
 

class KafkaConsumer():
    
    def __init__(self, broker, registry, topics, consumerID,avro=True,cb=0):    
        self.consumerID = consumerID
        self.broker = broker
        self.registry = registry
        self.topics = topics
        self.cb = cb
        self.avro = avro

        if(self.avro):
            self.consumerConf = {'bootstrap.servers': broker,'schema.registry.url': registry,'group.id': consumerID,'client.id': consumerID,'auto.offset.reset' : "earliest", 'max.poll.interval.ms' : 3000000}
            self.consumer = AvroConsumerStrKey(self.consumerConf)
        else:
            self.consumerConf = {'bootstrap.servers': broker,'group.id': consumerID,'client.id': consumerID,'auto.offset.reset' : "earliest", 'max.poll.interval.ms' : 3000000}
            self.consumer = Consumer(self.consumerConf)
        self.running = True
        
        self._subscribe(topics)        
        self.th = None

    def translate_to_topic_notation(edge_id):
        edge_id_copy = edge_id
        conform_edge_id = ""
        while edge_id_copy.find("---") != -1:
            conform_edge_id = conform_edge_id + edge_id_copy[:edge_id_copy.find("---")] + "---" + format(ord("-"), "x") + "---" + format(ord("-"), "x") + "---" + format(ord("-"), "x")
            edge_id_copy = edge_id_copy[edge_id_copy.find("---") + 3:]
    
        conform_edge_id = conform_edge_id + edge_id_copy
    
        return conform_edge_id.translate({ord(sign): "---" + format(ord(sign), "x") for sign in conform_edge_id if sign.isalnum() is False and sign != "-"})


    def run(self):
        self._subscribe(self.topics)
        while(self.running):
            try:
                inMsg = self.consumer.poll(10)
                if inMsg != None and not(inMsg.error()):
                    if(self.cb):
                        self.cb(inMsg)
                    else:
                        pass
                        # print("> > > received msg! topic =", inMsg.topic(), ", time =", inMsg.timestamp()[1],", key =", inMsg.key(), ", value =", inMsg.value(), flush=True)

            except SerializerError as e:
                print("Message deserialization failed for {}: {}".format(inMsg, e), flush=True)  
                
    def listenInBG(self):
       self.th = threading.Thread(target=self.run)
       self.th.start()
           

    # def listen(self):
    #     self.consumer.subscribe(self.topics)
    #     while(self.running):
    #         try:
    #             inMsg = self.consumer.poll(10)
    #             if inMsg != None and not(inMsg.error()):
    #                 if(self.cb):
    #                     self.cb(inMsg)
    #                 else:
    #                     print("> > > received msg! topic =", inMsg.topic(), ", time =", inMsg.timestamp()[1],", key =", inMsg.key(), ", value =", inMsg.value(), flush=True)

    #         except SerializerError as e:
    #             print("Message deserialization failed for {}: {}".format(inMsg, e), flush=True)



    def poll(self,timeout):
        try:
            msg = self.consumer.poll(timeout)
            if msg is None:
                return None
            if(msg.error()):
                
                print("got an error, restarting consumers")
                
                self.consumer.close()
                if(self.avro):
                    self.consumer = AvroConsumerStrKey(self.consumerConf)
                else:
                    self.consumer = Consumer(self.consumerConf)
                    
                self._subscribe(self.topics)   
                return None
            return msg
        except:
            print("got an exception, restarting consumers")
            self.consumer.close()
            if(self.avro):
                self.consumer = AvroConsumerStrKey(self.consumerConf)
            else:
                self.consumer = Consumer(self.consumerConf)
                
            self._subscribe(self.topics)   
            return None

    def _subscribe(self,topics):
        self.consumer.subscribe(topics)
        print("subscribed to:")
        for t in topics:        
            print(t)

    def stop(self):
        self.running = False
        self.running = False
        self.consumer.close()

                



